### PR TITLE
PHPC-2210 and PHPC-2211: Fix direct BSON copy and toJson memory leak

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/PHPC-2210
https://jira.mongodb.org/browse/PHPC-2211

### CODE REVIEW

File: tests/bulk/bulkwrite-insert-002.phpt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+<?php
+
+require_once __DIR__ . "/../utils/basic.inc";
+
+$manager = create_test_manager();
+
+$document = MongoDB\BSON\Document::fromJSON('{ "_id": 2 }');
+
+$bulk = new MongoDB\Driver\BulkWrite();
+$insertedId = $bulk->insert($document);
```
Comment (jmikola) [inline tests/bulk/bulkwrite-insert-002.phpt:17]: While implementing this test, I realized there's a separate bug where `_id` fields are not generated for Document instances. I opened [PHPC-2216](https://jira.mongodb.org/browse/PHPC-2216) to track that.

File: src/phongo_bson_encode.c [hunk 2]
```diff
@@ -403,17 +421,15 @@ static void php_phongo_zval_to_bson_internal(zval* data, php_phongo_field_path*
 			if (instanceof_function(Z_OBJCE_P(data), php_phongo_document_ce)) {
 				php_phongo_document_t* intern = Z_DOCUMENT_OBJ_P(data);
 
-				bson_destroy(bson);
-				bson_copy_to(intern->bson, bson);
+				phongo_bson_copy_to_noinit(intern->bson, bson);
```
Comment (jmikola) [inline src/phongo_bson_encode.c:424]: Note that only `php_phongo_zval_to_bson_internal` was affected.

While `php_phongo_bson_append_object` also handles direct copying of Document and PackedArray data, it does so via `bson_append_document` and `bson_append_array()` functions, respectively. There was no problem there.

Comment (alcaeus) [reply src/phongo_bson_encode.c:424]: Ah. I assumed that I could destroy the existing `bson_t` instance and have `bson_copy_to` handle the rest. If I understand this correctly, the issue is that `bson_destroy(bson)` frees the memory `bson` is pointing at. `bson_copy_to` doesn't catch this as it asserts on `dst`, which is still a valid `bson_t*` (i.e. not `NULL`) but pointing at memory no longer owned. Classic use of a footgun provided by C I have to admit, and a great catch on your part!

File: tests/query/query-ctor-007.phpt [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+$bulk->insert(['_id' => 1]);
+$bulk->insert(['_id' => 2]);
+$manager->executeBulkWrite(NS, $bulk);
+
+$filter = MongoDB\BSON\Document::fromJSON('{ "_id": { "$gt": 1 } }');
+$query = new MongoDB\Driver\Query($filter);
+
+var_dump($query);
+
+$cursor = $manager->executeQuery(NS, $query);
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply tests/query/query-ctor-007.phpt:24]: I agree - I believe the tests you added cover every aspect of potential document usage. 👍

**Review by alcaeus — APPROVED** [review]
LGTM. I've pushed a commit with clang-format to fix code style issues stemming from copying the logic over from libbson.

File: src/phongo_bson_encode.c (hunk 2 above)
Comment (jmikola) [reply src/phongo_bson_encode.c:424]: That's part of the problem, but removing `bson_destroy()` also doesn't help us. The crux of the matter is that `bson_copy_to()` isn't designed to work on an allocated `bson_t` struct, which always exist in an initialized state.

I spoke to @kevinAlbs about this and we think it's technically possible to alter `bson_copy_to()` to support both, but it doesn't seem necessary at this time -- and a local copy method in PHPC works just fine for our needs.

**Pull request merged**